src/support/allocators/node_allocator/memory_resource.h

```diff
@@ -0,0 +1,196 @@
+    return ((size_max + alignment_max - 1U) / alignment_max) * alignment_max;
+}
+
+/**
+ * Calculates dynamic memory usage of a MemoryResource. Not a member of MemoryResource so we can implement this
+ * in a cpp and don't depend on memusage.h in the header.
+ */
+[[nodiscard]] size_t DynamicMemoryUsage(size_t pool_size_bytes, std::vector<std::unique_ptr<char[]>> const& allocated_pools);
+
+} // namespace detail
+
+/**
+ * Actually holds and provides memory to an allocator. MemoryResource is an immobile object. It
+ * stores a number of memory pools which are used to quickly give out memory of a fixed allocation
+ * size. The class is purposely kept very simple. It only knows about "Allocations" and "Pools".
+ *
+ * - Pool: MemoryResource allocates one memory pool at a time. These pools are kept around until the
+ * memory resource is destroyed.
+ *
+ * - Allocations: Node-based containers allocate one node at a time. Whenever that happens, the
+ * MemoryResource's Allocate() gives out memory for one node. These are carved out from a previously
+ * allocated memory pool, or from a free list if it contains entries. Whenever a node is given back
+ * with Deallocate(), it is put into that free list.
+ */
+template <size_t ALLOCATION_SIZE_BYTES>
+class MemoryResource
+{
+    /**
+     * Size in bytes to allocate per pool, currently hardcoded to multiple of ALLOCATION_SIZE_BYTES
+     * that comes closest to 256 KiB.
+     */
+    static constexpr size_t POOL_SIZE_BYTES = (262144 / ALLOCATION_SIZE_BYTES) * ALLOCATION_SIZE_BYTES;
+
+public:
+    /**
+     * Construct a new MemoryResource object that uses the specified allocation size to optimize for.
+     */
+    MemoryResource() = default;
+
+    /**
+     * Copying/moving a memory resource is not allowed; it is an immobile object.
+     */
+    MemoryResource(const MemoryResource&) = delete;
+    MemoryResource& operator=(const MemoryResource&) = delete;
+    MemoryResource(MemoryResource&&) = delete;
+    MemoryResource& operator=(MemoryResource&&) = delete;
+
+    /**
+     * Deallocates all allocated pools.
+     *
+     * There's no Clear() method on purpose, because it would be dangerous. E.g. when calling
+     * clear() on an unordered_map, it is not certain that all allocated nodes are given back to the
+     * MemoryResource. Microsoft's STL still uses a control structure that might have the same size
+     * as the nodes, and therefore needs to be kept around until the map is actually destroyed.
+     */
+    ~MemoryResource() = default;
+
+    /**
+     * Allocates memory for sizeof(T).
+     *
+     * @tparam T Object to allocate memory for.
+     */
+    template <typename T>
+    [[nodiscard]] T* Allocate()
+    {
+        static_assert(ALLOCATION_SIZE_BYTES == detail::RequiredAllocationSizeBytes<T>());
+        static_assert(__STDCPP_DEFAULT_NEW_ALIGNMENT__ >= std::alignment_of_v<T>, "make sure AllocatePool() aligns correctly");
```

Comment:
I think that check is necessary. In `AllocatePool` I'm doing `new char[POOL_SIZE_BYTES]`, and this uses `__STDCPP_DEFAULT_NEW_ALIGNMENT__` as the alignment for the pool's memory. So if you try to use the node_allocator with a type that has an alignment > `__STDCPP_DEFAULT_NEW_ALIGNMENT__` this is a problem, even when `detail::RequiredAllocationSizeBytes` is done correctly. I had a test case that used 16 byte alignment, and with that `static_assert` that test failed on a 32bit machine.

Reply:
I think I'd prefer to keep the `static_assert` at the `Allocate` entry point, because currently `AllocatePool` doesn't need to know the type `T`. I guess it's more a matter of style :)